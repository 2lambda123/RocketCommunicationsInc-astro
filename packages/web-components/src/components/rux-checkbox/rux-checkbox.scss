:host {
    *,
    *::before,
    *::after {
        box-sizing: inherit;
    }

    display: inline-block;

    /** @prop --checkbox-label-color: [DEPRECATED] Label text color */
    --checkbox-label-color: var(--color-text-primary);

    /** @prop --checkbox-background-color: [DEPRECATED] Checkbox background color */
    --checkbox-background-color: var(--color-background-base-default);

    /** @prop --checkbox-border-color: [DEPRECATED] Checkbox border color */
    --checkbox-border-color: var(--color-border-interactive-muted);

    /** @prop --checkbox-checked-color: [DEPRECATED] Checkmark color when checked */
    --checkbox-checked-color: var(--color-background-interactive-default);

    /** @prop --checkbox-hover-border-color: [DEPRECATED] Hover border color */
    --checkbox-hover-border-color: var(--color-background-interactive-hover);
}

.hidden,
:host([hidden]) {
    display: none;
}

.rux-form-field {
    display: flex;
    flex-direction: column;
    font-family: var(--font-body-1-font-family);
    font-size: var(--font-body-1-font-size);
    letter-spacing: var(--font-body-1-letter-spacing);
    font-weight: var(--font-body-1-font-weight);
    color: var(--color-text-primary);
}

.rux-checkbox {
    display: flex;
    position: relative;
    line-height: 1.2;

    &--indeterminate {
        input[type='checkbox'] + label::after {
            position: absolute;
            display: flex;
            content: '';
            top: 5px;
            width: 10px;
            height: 5px;
            transform: rotate(0deg);
            border-right: 0;
            border-top: 0;
            border-bottom: 2px solid var(--color-background-interactive-default);
            left: 4px;
        }
    }

    input[type='checkbox'] {
        -webkit-appearance: none;
        appearance: none;

        + label {
            position: relative;
            display: flex;
            align-items: center;
            justify-content: flex-start;
            color: var(--color-text-primary);
            letter-spacing: 0.5px;
            cursor: pointer;
            margin-left: -7px;

            // Box
            &::before {
                display: flex;
                content: '';
                align-self: start;
                height: 1.125rem;
                width: 1.125rem;

                margin: 0 0.625rem 0 0;
                background-color: var(--color-background-base-default);
                border: 1px solid var(--color-border-interactive-muted);
                border-radius: var(--radius-checkbox);
            }
        }
        //If no label is passed in we don't want the margin
        + .rux-checkbox--no-label {
            &::before {
                margin: 0;
            }
        }

        // checked
        &:checked {
            + label {
                &::after {
                    position: absolute;
                    display: flex;
                    content: '';
                    top: 5px;
                    height: 6px;
                    width: 12px;
                    left: 3px;
                    border-right: 2px solid
                        var(--color-background-interactive-default);
                    border-top: 2px solid
                        var(--color-background-interactive-default);
                    transform: rotate(125deg);
                }
                &::before {
                    border-color: var(--color-border-interactive-muted);
                }
            }
        }

        &:disabled {
            + label {
                cursor: not-allowed;
                opacity: var(--disabled-opacity);
            }
        }
    }

    input[type='checkbox']:not(:disabled):hover {
        + label {
            &::before {
                border-color: var(--color-background-interactive-hover);
            }
        }
    }

    &--has-text {
        margin-bottom: 0;
    }

    + .rux-help-text {
        margin-bottom: 0.75rem;
    }
}

.rux-checkbox-label__asterisk {
    margin-left: 4px;
}

.rux-help-text {
    margin-top: 0.625rem;
    color: var(--color-text-primary);
    font-size: var(--font-body-2-font-size);
    font-family: var(--font-body-2-font-family);
    font-weight: var(--font-body-2-font-weight);
    letter-spacing: var(--font-body-2-letter-spacing);
}
