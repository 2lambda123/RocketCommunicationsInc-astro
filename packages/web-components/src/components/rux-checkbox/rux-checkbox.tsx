import {
    Component,
    h,
    Prop,
    Event,
    EventEmitter,
    Element,
    Watch,
    Host,
    State,
} from '@stencil/core'
import FormFieldMessage from '../../common/functional-components/FormFieldMessage/FormFieldMessage'

import { FormFieldInterface } from '../../common/interfaces.module'
import { renderHiddenInput, hasSlot } from '../../utils/utils'

let id = 0

/**
 * @slot (default) - the label of the checkbox.
 * @part form-field - the form field wrapper container
 * @part help-text - The help text element
 * @part label - the label of rux-checkbox
 */
@Component({
    tag: 'rux-checkbox',
    styleUrl: 'rux-checkbox.scss',
    shadow: true,
})
export class RuxCheckbox implements FormFieldInterface {
    private checkboxId = `rux-checkbox-${++id}`
    private _inputEl?: HTMLInputElement

    @Element() el!: HTMLRuxCheckboxElement

    @State() hasLabelSlot = false

    /**
     * The help or explanation text
     */
    @Prop({ attribute: 'help-text' }) helpText?: string

    /**
     * The checkbox name
     */
    @Prop() name = ''

    /**
     * The checkbox value
     */
    @Prop({ reflect: true, mutable: true }) value: string = ''

    /**
     * The checkbox label text. For HTML content, use the default slot instead.
     */
    @Prop() label?: string

    /**
     * Toggles checked state of a checkbox
     */
    @Prop({ reflect: true, mutable: true }) checked: boolean = false

    @Watch('checked')
    updateChecked() {
        if (this._inputEl) {
            this._inputEl.checked = this.checked
        }
    }

    /**
     * Toggles indeterminate state of a checkbox. The indeterminate property does not exist in HTML, but can be set in JS. [HTML Checkbox & Indeterminate State](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#indeterminate)
     */
    @Prop({ reflect: true, mutable: true }) indeterminate: boolean = false

    @Watch('indeterminate')
    updateIndeterminate() {
        if (this._inputEl) {
            this._inputEl.indeterminate = this.indeterminate
        }
    }

    /**
     * Disables the checkbox via HTML disabled attribute. Checkbox takes on a distinct visual state. Cursor uses the not-allowed system replacement and all keyboard and mouse events are ignored.
     */
    @Prop({ reflect: true }) disabled: boolean = false

    /**
     * Fired when the value of the input changes - [HTMLElement/input_event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event)
     */
    @Event({ eventName: 'ruxchange' }) ruxChange!: EventEmitter

    /**
     * Fired when an alteration to the input's value is committed by the user - [HTMLElement/change_event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event)
     */
    @Event({ eventName: 'ruxinput' }) ruxInput!: EventEmitter

    /**
     * Fired when an element has lost focus - [HTMLElement/blur_event](https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event)
     */
    @Event({ eventName: 'ruxblur' }) ruxBlur!: EventEmitter

    connectedCallback() {
        this._onClick = this._onClick.bind(this)
        this._onInput = this._onInput.bind(this)
        this._checkForLabelSlot = this._checkForLabelSlot.bind(this)
    }

    componentWillLoad() {
        this._checkForLabelSlot()
    }

    componentDidLoad() {
        if (this._inputEl && this.indeterminate) {
            // indeterminate property does not exist in HTML but is accessible via js
            this._inputEl.indeterminate = true
        }
    }

    get hasLabel() {
        return this.label ? true : this.hasLabelSlot
    }

    private _checkForLabelSlot() {
        this.hasLabelSlot = hasSlot(this.el)
    }

    private _onClick(e: Event): void {
        const target = e.target as HTMLInputElement
        if (this.indeterminate) {
            this.indeterminate = false
        }
        this.checked = target.checked
        this.ruxChange.emit()
    }

    private _onInput(e: Event) {
        const target = e.target as HTMLInputElement
        this.value = target.value
        this.ruxInput.emit()
    }

    private _onBlur = () => {
        this.ruxBlur.emit()
    }

    render() {
        const {
            _checkForLabelSlot,
            _onBlur,
            _onClick,
            _onInput,
            checkboxId,
            checked,
            disabled,
            el,
            helpText,
            name,
            value,
            indeterminate,
            label,
            hasLabel,
            hasLabelSlot,
        } = this

        if (!indeterminate) {
            renderHiddenInput(true, el, name, value || 'on', disabled, checked)
        }

        return (
            <Host>
                <div class="rux-form-field" part="form-field">
                    <label
                        class={{
                            'rux-checkbox': true,
                            'rux-checkbox--disabled': disabled,
                        }}
                        htmlFor={checkboxId}
                    >
                        <input
                            type="checkbox"
                            class={{
                                'rux-checkbox__input': true,
                                'rux-checkbox__input--no-label': !hasLabel,
                            }}
                            name={name}
                            id={checkboxId}
                            disabled={disabled}
                            checked={checked}
                            //Allows storybook's indetermiante control to take effect.
                            indeterminate={indeterminate}
                            value={value}
                            onChange={_onClick}
                            onInput={_onInput}
                            onBlur={_onBlur}
                            ref={(el) => (this._inputEl = el)}
                        />
                        <div class="rux-checkbox__control">
                            {indeterminate ? (
                                <svg
                                    width="18"
                                    height="18"
                                    fill="none"
                                    xmlns="http://www.w3.org/2000/svg"
                                    viewBox="0 0 18 18"
                                >
                                    <rect
                                        x=".5"
                                        y=".5"
                                        width="17"
                                        height="17"
                                        rx="1.5"
                                        fill="var(--_checkbox-color-background)"
                                    />
                                    <path
                                        fill="var(--_checkbox-checkmark-color-fill)"
                                        d="M4 8h10v2H4z"
                                    />
                                    <rect
                                        x=".5"
                                        y=".5"
                                        width="17"
                                        height="17"
                                        rx="1.5"
                                        stroke="var(--_checkbox-color-border)"
                                    />
                                </svg>
                            ) : checked ? (
                                <svg
                                    width="18"
                                    height="18"
                                    fill="none"
                                    xmlns="http://www.w3.org/2000/svg"
                                    viewBox="0 0 18 18"
                                >
                                    <rect
                                        x=".5"
                                        y=".5"
                                        width="17"
                                        height="17"
                                        rx="1.5"
                                        fill="var(--_checkbox-color-background)"
                                    />
                                    <path
                                        fill="var(--_checkbox-checkmark-color-fill)"
                                        d="m14.899 4.806-6.883 9.83-1.639-1.147 6.883-9.83z"
                                    />
                                    <path
                                        fill="var(--_checkbox-checkmark-color-fill)"
                                        d="m9.163 12.997-1.147 1.638L3.1 11.194l1.147-1.638z"
                                    />
                                    <rect
                                        x=".5"
                                        y=".5"
                                        width="17"
                                        height="17"
                                        rx="1.5"
                                        stroke="var(--_checkbox-color-border)"
                                    />
                                </svg>
                            ) : (
                                <svg
                                    width="18"
                                    height="18"
                                    xmlns="http://www.w3.org/2000/svg"
                                    fill="none"
                                    viewBox="0 0 18 18"
                                >
                                    <rect
                                        x=".5"
                                        y=".5"
                                        width="17"
                                        height="17"
                                        rx="1.5"
                                        fill="var(--_checkbox-color-background)"
                                    />
                                    <rect
                                        x=".5"
                                        y=".5"
                                        width="17"
                                        height="17"
                                        rx="1.5"
                                        stroke="var(--_checkbox-color-border)"
                                    />
                                </svg>
                            )}
                        </div>
                        <div
                            part="label"
                            class={{
                                'rux-checkbox__label': true,
                                hidden: !hasLabel,
                            }}
                        >
                            {hasLabelSlot ? null : label}
                            <slot onSlotchange={_checkForLabelSlot} />
                        </div>
                    </label>
                </div>
                <FormFieldMessage helpText={helpText} />
            </Host>
        )
    }
}
