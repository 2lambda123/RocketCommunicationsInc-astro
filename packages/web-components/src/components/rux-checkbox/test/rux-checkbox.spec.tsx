import { newSpecPage } from '@stencil/core/testing'
import { RuxCheckbox } from '../rux-checkbox'

describe('rux-checkbox', () => {
    it('builds', async () => {
        const checkbox = new RuxCheckbox()
        const {
            checkboxId,
            checked,
            disabled,
            indeterminate,
            name,
            value,
        } = checkbox

        expect(checkbox).toBeTruthy()
        expect({
            checkboxId,
            checked,
            disabled,
            indeterminate,
            name,
            value,
        }).toEqual({
            checkboxId: 'rux-checkbox-1',
            checked: false,
            disabled: false,
            errorText: undefined,
            indeterminate: false,
            name: '',
            value: '',
        })
    })

    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxCheckbox],
            html: `<rux-checkbox></rux-checkbox>`,
        })
        expect(page.root).toMatchSnapshot()
    })
})
