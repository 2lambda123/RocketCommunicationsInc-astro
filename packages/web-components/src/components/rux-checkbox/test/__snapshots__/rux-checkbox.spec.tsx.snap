// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-checkbox renders 1`] = `
<rux-checkbox value="">
  <mock:shadow-root>
    <div class="rux-form-field" part="form-field">
      <div class="rux-checkbox">
        <input class="rux-checkbox__input rux-checkbox__input--no-label" id="rux-checkbox-2" type="checkbox" value="">
        <label htmlfor="rux-checkbox-2" part="label">
          <span class="hidden">
            <slot></slot>
          </span>
        </label>
      </div>
    </div>
  </mock:shadow-root>
</rux-checkbox>
`;
